cuML/src/common/device_buffer.hpp

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "../../../ml-prims/src/utils.h"
+
+#include "../cuML.hpp"
+
+namespace ML {
+
+/**
+ * RAII object owning a contigous typed device buffer. The passed in allocator supports asynchronus allocation and
+ * deallocation so this can be used for temporary memory 
+ * @code{.cpp}
+ * void foo( cumlHandle* handle, .., cudaStream_t stream )
+ * {
+ *     ...
+ *     device_buffer<T> temp( handle->getHostAllocator(), 0 )
+ *     
+ *     temp.resize(n, stream);
+ *     kernelA<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     kernelB<<<grid,block,0,stream>>>(...,temp.data(),...);
+ *     temp.release(stream);
+ * }
+ * @endcode
+ */
+template<typename T>
+class device_buffer
+{
+public:
+    typedef std::size_t         size_type;
+    typedef T                   value_type;
+    typedef value_type*         iterator;
+    typedef const value_type*   const_iterator;
+    typedef T&                  reference;
+    typedef const T&            const_reference;
+
+    device_buffer() = delete;
+
+    device_buffer(const device_buffer& other) = delete;
+
+    device_buffer& operator=(const device_buffer& other) = delete;
+
+    device_buffer(std::shared_ptr<deviceAllocator> allocator, size_type n = 0)
```

Comment:
It's a bit cumbersome to always have to pass an allocator as a parameter whenever I want to do an allocation with RAII.

It'd be nicer if it could be a template parameter such that you could have an alias for a `device_buffer` that always uses `my_allocator`.

Reply:
I have thought about that quite a bit. The advantage of passing in an allocator as a runtime argument over a template parameter are: 
* Irrespective of the used allocator the containers remain type compatible.
* We can easily have multiple different allocator objects, e.g. one per device one per thread, .... With a template argument we would need to reference a global object. 
Does that make sense to you?